tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
+  return rewriter.create<mhlo::DynamicReshapeOp>(loc, outTy, tensor, mhloShape)
+      .getResult();
+}
+
+// Converts DotGeneralOp if it represents vec * vec, vec * mat, mat * vec.
+// Will first unsqueeze vector to matirx, then convert DoGeneralOp to
+// matrix multiply; After that drop(unsqueeze) the extra-dimensions from
+// the output of DoGeneralOp.
+struct DotGeneralConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
+  explicit DotGeneralConvert(MLIRContext* context)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter& rewriter) const override {
+    Location loc = op.getLoc();
+    Value old_lhs = op.lhs();
+    Value old_rhs = op.rhs();
+
+    RankedTensorType old_l_type =
+        old_lhs.getType().dyn_cast<RankedTensorType>();
+    RankedTensorType old_r_type =
+        old_rhs.getType().dyn_cast<RankedTensorType>();
+    if ((!old_l_type || !old_r_type)) {
+      return failure();
+    }
+
+    auto dim_numbers = op.dot_dimension_numbers();
+    auto batching_dims = dim_numbers.getLhsBatchingDimensions();
+    auto n_batch_dims = batching_dims.size();
+    for (auto d : batching_dims)
+      if (d >= n_batch_dims)
+        return rewriter.notifyMatchFailure(
+            op, "the batching_dims must be the leading dimensions");
+
+    auto lhs_contracting_dims = dim_numbers.getLhsContractingDimensions();
+    auto rhs_contracting_dims = dim_numbers.getRhsContractingDimensions();
+    if (lhs_contracting_dims.size() != 1 || rhs_contracting_dims.size() != 1)
+      return rewriter.notifyMatchFailure(op, "multiple contracting dimensions");
+
+    size_t old_lhs_rank = old_l_type.getRank();
+    size_t old_rhs_rank = old_r_type.getRank();
+    Value new_lhs = old_lhs;
+    Value new_rhs = old_rhs;
+
+    auto old_out_ty = op.getType().dyn_cast<RankedTensorType>();
+    auto old_out_shape = old_out_ty.getShape();
+    SmallVector<int64_t, 4> new_out_shape{old_out_shape.begin(),
+                                          old_out_shape.begin() + n_batch_dims};
+    while (new_out_shape.size() < n_batch_dims + 2) new_out_shape.push_back(-1);
+
+    auto old_lhs_matrix_rank = old_lhs_rank - n_batch_dims;
+    auto old_rhs_matrix_rank = old_rhs_rank - n_batch_dims;
+    // Only do conversion for mhlo::DotGeneralOp lowering from mhlo::DotOp,
+    // mhlo::EinsumOp can be convert similar, can be added if it's needed.
+    //
+    // unsqueeze lhs & rhs to matrix if needed
+    if (n_batch_dims == lhs_contracting_dims[0] && old_lhs_matrix_rank == 1) {
+      if (old_rhs_matrix_rank == 1) {
+        // vec x vec
+        new_lhs = unsqueezeTensorDim(rewriter, op, old_lhs, n_batch_dims);
+        new_rhs = unsqueezeTensorDim(rewriter, op, old_rhs, n_batch_dims + 1);
+        new_out_shape[n_batch_dims] = 1;
+        new_out_shape[n_batch_dims + 1] = 1;
+      } else if (n_batch_dims == lhs_contracting_dims[0]) {
```

Comment:
better to use a assert, instead of a `if` statement.

Reply:
Actually there is a bug before.